### PR TITLE
SOLR-13677 fix & cleanup

### PR DESCRIPTION
**Pull request by sigram**
# Description

This PR is based on Noble's patch plus cleanups & refactoring identified in the Jira review.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests
`MetricsHandlerTest.testMetricsUnload()` is probably insufficient to test all corner cases. We still need to add tests for hierarchical metric tags and back-compat of the API.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I am authorized to contribute this code to the ASF and have removed any code I do not have a license to distribute.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [ ] I have run `ant precommit` and the appropriate test suite.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java [hunk 1]
```diff
@@ -431,12 +432,13 @@ public void register() {
       cache.setState(SolrCache.State.LIVE);
       infoRegistry.put(cache.name(), cache);
     }
-    metricManager = core.getCoreContainer().getMetricManager();
-    registryName = core.getCoreMetricManager().getRegistryName();
+    this.solrMetricsContext = core.getSolrMetricsContext().getChildContext(this);
```
Comment (sigram) [inline solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java]: `SolrIndexSearcher` is not a `SolrMetricProducer` but it registers some gauges so I think it needs a child context - especially since its life-cycle is different from that of SolrCore.

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java [hunk 2]
```diff
@@ -108,7 +108,7 @@
   
   private boolean isV2Enabled = !"true".equals(System.getProperty("disable.v2.api", "false"));
 
-  private final String metricTag = Integer.toHexString(hashCode());
+  private final String metricTag = getClass().getSimpleName() + "@" + Integer.toHexString(hashCode());
```
Comment (sigram) [inline solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java]: Should use the new method from `SolrMetricProducer`.

File: solr/core/src/java/org/apache/solr/update/SolrIndexWriter.java [hunk 3]
```diff
@@ -135,8 +135,7 @@ private SolrIndexWriter(SolrCore core, String name, String path, Directory direc
     infoStream = getConfig().getInfoStream();
     this.directory = directory;
     numOpens.incrementAndGet();
-    metricManager = core.getCoreContainer().getMetricManager();
-    registryName = core.getCoreMetricManager().getRegistryName();
+    solrMetricsContext = core.getSolrMetricsContext().getChildContext(this);
```
Comment (sigram) [inline solr/core/src/java/org/apache/solr/update/SolrIndexWriter.java]: Similar situation to `SolrIndexSearcher`, the life-cycle here is different from SolrCore.

File: solr/core/src/java/org/apache/solr/update/SolrIndexWriter.java [hunk 4]
```diff
@@ -345,6 +344,9 @@ private void cleanup() throws IOException {
       if (directoryFactory != null) {
         directoryFactory.release(directory);
       }
+      if (solrMetricsContext != null) {
```
Comment (sigram) [inline solr/core/src/java/org/apache/solr/update/SolrIndexWriter.java]: Not a `SolrMetricProducer` so we need to call this explicitly.

File: solr/core/src/test/org/apache/solr/handler/admin/MetricsHandlerTest.java [hunk 5]
```diff
@@ -325,8 +335,123 @@ public void testKeyMetrics() throws Exception {
...
         MetricsHandler.KEY_PARAM, "solr.jetty:unknown:baz"), resp);
     values = resp.getValues();
-    metrics = (NamedList)values.get("metrics");
+    metrics = (NamedList) values.get("metrics");
     assertEquals(0, metrics.size());
     assertNotNull(values.findRecursive("errors", "solr.jetty:unknown:baz"));
   }
+
+  @Test
+  public void testMetricsUnload() throws Exception {
```
Comment (sigram) [inline solr/core/src/test/org/apache/solr/handler/admin/MetricsHandlerTest.java]: We need another test that tests the following scenario:
* instance 1 registers its gauges
* instance 2 is created and registers its gauges
* instance 1 is closed and should unregister its gauges

together with dependent child instances

File: solr/core/src/java/org/apache/solr/security/AuthenticationPlugin.java [hunk 6]
```diff
@@ -184,7 +181,7 @@ public Category getCategory() {
 
   @Override
   public MetricRegistry getMetricRegistry() {
-    return registry;
+    return metrics == null ? null : metrics.getMetricRegistry();
```
[the remark this answers is not in the document]
Comment (sigram) [reply solr/core/src/java/org/apache/solr/security/AuthenticationPlugin.java]: Good point, we can.

**Comment by sigram** [comment]
Regarding scope - in many places the SolrMetricsContext instance can be created once in the parent and then used for many components that differ only in their scope. Passing one common instance + component-specific scope seemed more efficient than creating new component-specific context for each component just because the scope was different ... but I'm not strongly attached to this change if you have strong arguments otherwise.

File: solr/core/src/java/org/apache/solr/security/AuthenticationPlugin.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sigram) [reply solr/core/src/java/org/apache/solr/security/AuthenticationPlugin.java]: Actually, we can't because we already have a default with the same signature in `SolrInfoBean`, but I can improve the implementation there to avoid this repetition.

File: solr/core/src/java/org/apache/solr/metrics/SolrMetricProducer.java [hunk 7]
```diff
@@ -19,17 +19,72 @@
...
+
+  @Override
+  default void close() throws Exception {
+    SolrMetricsContext context = getSolrMetricsContext();
+    if (context == null) {
+      return;
+    } else {
+      context.unregister();
+    }
+    // ??? (ab) no idea what this was supposed to avoid
```
[the remark this answers is not in the document]
Comment (sigram) [reply solr/core/src/java/org/apache/solr/metrics/SolrMetricProducer.java]: See my reply below. Still, regarding this particular snippet, I don't think it's needed anymore,, maybe it's a leftover from an earlier refactoring?

**Comment by sigram** [comment]
Committed.
